File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def solution(today, terms, privacies):
    ty, tm, td = map(int, today.split('.'))
    
    term_dict = dict()
    answer = []
    
    for term in terms:
        key, value = term.split()
        term_dict[key] = int(value) * 28
    
    for i in range(len(privacies)):
        privacy = privacies[i]
        date, term = privacy.split()
        y, m, d = map(int, date.split('.'))
        
        d += term_dict[term] - 1
        
        # day -> extra month + day로 조정
        extra_m = d // 28
        m += extra_m - 1 if d % 28 == 0 else extra_m
        d = 28 if d % 28 == 0 else d % 28
        
        # month -> extra year + month로 조정
        extra_y = m // 12
        y += extra_y - 1 if m % 12 == 0 else extra_y
        m = 12 if m % 12 == 0 else m % 12
        
        if y < ty:
            answer.append(i + 1)
        elif y == ty:
            if m < tm:
                answer.append(i + 1)
            elif m == tm:
                if d < td:
                    answer.append(i + 1)
        
    
    return answer

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.js
function solution(today, terms, privacies) {
    const answer = [];
    const term_rules = new Map();
    const [y, m, d] = today.split(".").map((x) => Number(x));
    const current = (y - 1) * 12 * 28 + (m - 1) * 28 + d;
    
    for (const term of terms) {
        const [t, m] = term.split(" ");
        term_rules.set(t, m * 28);
    }
    
    let idx = 1;
    
    for (const privacy of privacies) {
        const [date, type] = privacy.split(" ")
        const [y, m, d] = date.split(".").map((x) => Number(x));
        
        const limit = (y - 1) * 12 * 28 + (m - 1) * 28 + d + term_rules.get(type);
        if (current >= limit) answer.push(idx);
        idx++;
    }

    return answer;
}